internal/datastore/postgres/readwrite.go

```diff
@@ -0,0 +1,258 @@
+	}
+	if filter.OptionalRelation != "" {
+		query = query.Where(sq.Eq{colRelation: filter.OptionalRelation})
+		tracerAttributes = append(tracerAttributes, common.ObjRelationNameKey.String(filter.OptionalRelation))
+	}
+
+	// Add clauses for the SubjectFilter
+	if subjectFilter := filter.OptionalSubjectFilter; subjectFilter != nil {
+		query = query.Where(sq.Eq{colUsersetNamespace: subjectFilter.SubjectType})
+		tracerAttributes = append(tracerAttributes, common.SubNamespaceNameKey.String(subjectFilter.SubjectType))
+		if subjectFilter.OptionalSubjectId != "" {
+			query = query.Where(sq.Eq{colUsersetObjectID: subjectFilter.OptionalSubjectId})
+			tracerAttributes = append(tracerAttributes, common.SubObjectIDKey.String(subjectFilter.OptionalSubjectId))
+		}
+		if relationFilter := subjectFilter.OptionalRelation; relationFilter != nil {
+			query = query.Where(sq.Eq{colUsersetRelation: stringz.DefaultEmpty(relationFilter.Relation, datastore.Ellipsis)})
+			tracerAttributes = append(tracerAttributes, common.SubRelationNameKey.String(relationFilter.Relation))
+		}
+	}
+
+	span.SetAttributes(tracerAttributes...)
+
+	query = query.Set(colDeletedTxn, rwt.newTxnID)
+
+	sql, args, err := query.ToSql()
+	if err != nil {
+		return fmt.Errorf(errUnableToDeleteRelationships, err)
+	}
+
+	if _, err := rwt.tx.Exec(ctx, sql, args...); err != nil {
+		return fmt.Errorf(errUnableToDeleteRelationships, err)
+	}
+
+	return nil
+}
+
+func (rwt *pgReadWriteTXN) WriteNamespaces(newConfigs ...*core.NamespaceDefinition) error {
+	ctx := datastore.SeparateContextWithTracing(rwt.ctx)
+
+	ctx, span := tracer.Start(ctx, "WriteNamespaces")
+	defer span.End()
+
+	// span.SetAttributes(common.ObjNamespaceNameKey.String(newConfig.Name))
+
+	deletedNamespaceClause := sq.Or{}
+	writeQuery := writeNamespace
+
+	for _, newNamespace := range newConfigs {
+		serialized, err := proto.Marshal(newNamespace)
+		if err != nil {
+			return fmt.Errorf(errUnableToWriteConfig, err)
+		}
+		span.AddEvent("Serialized namespace config")
+
+		deletedNamespaceClause = append(deletedNamespaceClause, sq.Eq{colNamespace: newNamespace.Name})
+		writeQuery = writeQuery.Values(newNamespace.Name, serialized, rwt.newTxnID)
+	}
+
+	delSQL, delArgs, err := deleteNamespace.
+		Set(colDeletedTxn, rwt.newTxnID).
+		Where(sq.And{sq.Eq{colDeletedTxn: liveDeletedTxnID}, deletedNamespaceClause}).
+		ToSql()
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteConfig, err)
+	}
+
+	_, err = rwt.tx.Exec(ctx, delSQL, delArgs...)
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteConfig, err)
+	}
+
+	sql, args, err := writeQuery.ToSql()
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteConfig, err)
+	}
+
+	_, err = rwt.tx.Exec(ctx, sql, args...)
+	if err != nil {
+		return fmt.Errorf(errUnableToWriteConfig, err)
+	}
+	span.AddEvent("Namespace config written")
+
+	return nil
+}
+
+func (rwt *pgReadWriteTXN) DeleteNamespace(nsName string) error {
+	ctx := datastore.SeparateContextWithTracing(rwt.ctx)
```

Comment:
probably want to start a span here, too

Reply:
Added it. I would like to hold off on any unnecessary refactoring.